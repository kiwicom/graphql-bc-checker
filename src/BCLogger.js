// @flow

import os from 'os';

const breakingChangeMarkerStart = '<BREAKING-CHANGES-LOG>';
const breakingChangeMarkerEnd = '</BREAKING-CHANGES-LOG>';

type BreakingChange = {
  +type: string,
  +description: string,
  ...
};

const getBreakingChangesLog = (
  oldSnapshot: string,
  newBreakingChanges: ?$ReadOnlyArray<BreakingChange>,
) => {
  const breakingChangeBlockRegExp = new RegExp(
    `# ${breakingChangeMarkerStart}([\\s\\S]+)# ${breakingChangeMarkerEnd}`,
  );

  const match = oldSnapshot.match(breakingChangeBlockRegExp);
  let breakingChangesLog = match === null ? os.EOL : match[1];

  if (newBreakingChanges) {
    breakingChangesLog += newBreakingChanges.reduce(
      (acc, curVal) => `${acc}#  ${curVal.type}: ${curVal.description}${os.EOL}`,
      '',
    );
  }

  return breakingChangesLog;
};

export const buildBreakingChangesBlock = (
  oldSnapshot: string,
  newBreakingChanges: ?$ReadOnlyArray<BreakingChange>,
) => {
  const breakingChangesLog = getBreakingChangesLog(oldSnapshot, newBreakingChanges);

  return `# ${breakingChangeMarkerStart}${breakingChangesLog}# ${breakingChangeMarkerEnd}`;
};
