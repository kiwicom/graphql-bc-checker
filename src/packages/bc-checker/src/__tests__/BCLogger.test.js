// @flow

import { generateTestsFromFixtures } from '@kiwicom/test-utils';

import { buildBreakingChangesBlock } from '../BCLogger';

function operation(newBreakingChanges) {
  return input => buildBreakingChangesBlock(input, newBreakingChanges);
}

generateTestsFromFixtures(`${__dirname}/__fixtures__`, operation());
generateTestsFromFixtures(
  `${__dirname}/__fixtures__`,
  operation([
    {
      type: 'append',
      description: 'this',
    },
    {
      type: 'and append',
      description: 'also this',
    },
  ]),
);
