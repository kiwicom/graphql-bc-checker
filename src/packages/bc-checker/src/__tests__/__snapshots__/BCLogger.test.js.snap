// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches expected output: empty-block.txt 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
# <BREAKING-CHANGES-LOG>
# </BREAKING-CHANGES-LOG>

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
# <BREAKING-CHANGES-LOG>
# </BREAKING-CHANGES-LOG>
`;

exports[`matches expected output: empty-block.txt 2`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
# <BREAKING-CHANGES-LOG>
# </BREAKING-CHANGES-LOG>

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
# <BREAKING-CHANGES-LOG>
#  append: this
#  and append: also this
# </BREAKING-CHANGES-LOG>
`;

exports[`matches expected output: full-block.txt 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
# <BREAKING-CHANGES-LOG>
#  FIELD_REMOVED: Booking.unstable_baggage was removed.
#  FIELD_REMOVED: BookingOneWay.unstable_baggage was removed.
#  FIELD_REMOVED: BookingInterface.unstable_baggage was removed.
#  FIELD_REMOVED: BookingReturn.unstable_baggage was removed.
#  FIELD_REMOVED: BookingMulticity.unstable_baggage was removed.
# </BREAKING-CHANGES-LOG>

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
# <BREAKING-CHANGES-LOG>
#  FIELD_REMOVED: Booking.unstable_baggage was removed.
#  FIELD_REMOVED: BookingOneWay.unstable_baggage was removed.
#  FIELD_REMOVED: BookingInterface.unstable_baggage was removed.
#  FIELD_REMOVED: BookingReturn.unstable_baggage was removed.
#  FIELD_REMOVED: BookingMulticity.unstable_baggage was removed.
# </BREAKING-CHANGES-LOG>
`;

exports[`matches expected output: full-block.txt 2`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
# <BREAKING-CHANGES-LOG>
#  FIELD_REMOVED: Booking.unstable_baggage was removed.
#  FIELD_REMOVED: BookingOneWay.unstable_baggage was removed.
#  FIELD_REMOVED: BookingInterface.unstable_baggage was removed.
#  FIELD_REMOVED: BookingReturn.unstable_baggage was removed.
#  FIELD_REMOVED: BookingMulticity.unstable_baggage was removed.
# </BREAKING-CHANGES-LOG>

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
# <BREAKING-CHANGES-LOG>
#  FIELD_REMOVED: Booking.unstable_baggage was removed.
#  FIELD_REMOVED: BookingOneWay.unstable_baggage was removed.
#  FIELD_REMOVED: BookingInterface.unstable_baggage was removed.
#  FIELD_REMOVED: BookingReturn.unstable_baggage was removed.
#  FIELD_REMOVED: BookingMulticity.unstable_baggage was removed.
#  append: this
#  and append: also this
# </BREAKING-CHANGES-LOG>
`;

exports[`matches expected output: missing-block.txt 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
# <BREAKING-CHANGES-LOG>
# </BREAKING-CHANGES-LOG>
`;

exports[`matches expected output: missing-block.txt 2`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
# <BREAKING-CHANGES-LOG>
#  append: this
#  and append: also this
# </BREAKING-CHANGES-LOG>
`;
